File: backend/src/main/java/com/backend/petplace/domain/pet/entity/Pet.java
package com.backend.petplace.domain.pet.entity;

import com.backend.petplace.domain.pet.dto.request.CreatePetRequest;
import com.backend.petplace.domain.pet.dto.request.UpdatePetRequest;
import com.backend.petplace.domain.user.entity.User;
import com.backend.petplace.global.entity.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import java.time.LocalDate;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Pet extends BaseEntity {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "petId")
  private Long id;

  @ManyToOne
  @JoinColumn(name = "userId")
  private User user;

  @Column(name = "name")
  private String name;

  @Column(name = "gender", nullable = false)
  @Enumerated(EnumType.STRING)
  private Gender gender;

  @Column(name = "birthDate")
  private LocalDate birthDate;

  @Column(name = "type")
  private String type;

  @Builder
  public Pet(User user, String name, Gender gender, LocalDate birthDate, String type) {
    this.user = user;
    this.name = name;
    this.gender = gender;
    this.birthDate = birthDate;
    this.type = type;
  }

  public static Pet createPet(User user, CreatePetRequest request){
    return Pet.builder()
        .user(user)
        .name(request.getName())
        .gender(Gender.valueOf(request.getGender()))
        .birthDate(request.getBirthDate())
        .type(request.getType())
        .build();
  }

  public void assignUser(User user){ //엔티티 생성 이후 한번씩 등록해주기 (연관관계 편의 메서드)
    if(this.user != null){
      user.getPets().remove(this);
    }
    this.user = user; //user FK 적용
    user.getPets().add(this); //반대쪽에도 똑같이 동기화
  }

  public void updatePet(UpdatePetRequest request){ //업데이트 (선택적 업데이트)
    this.name = !request.isValidName() ? this.name : request.getName();
    this.gender = request.getGender() == null ? this.gender : Gender.valueOf(request.getGender());
    this.type = request.getType() == null ? this.type : request.getType();
    this.birthDate = request.getBirthDate() == null ? this.birthDate : request.getBirthDate();
  }

}

File: backend/src/main/java/com/backend/petplace/domain/pet/entity/Pet.kt
package com.backend.petplace.domain.pet.entity

import com.backend.petplace.domain.pet.dto.request.CreatePetRequest
import com.backend.petplace.domain.pet.dto.request.UpdatePetRequest
import com.backend.petplace.domain.user.entity.User
import com.backend.petplace.global.entity.BaseEntity
import jakarta.persistence.*
import java.time.LocalDate

@Entity
class Pet (

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "petId")
    val id: Long? = null,

    @ManyToOne
    @JoinColumn(name = "userId")
    var user: User? = null,

    @Column(name = "name")
    var name: String? = null,

    @Column(name = "gender", nullable = false)
    @Enumerated(EnumType.STRING)
    var gender: Gender? = null,

    @Column(name = "birthDate")
    var birthDate: LocalDate? = null,

    @Column(name = "type")
    var type: String? = null

) : BaseEntity() {

    companion object { //똑같은 객체 타입에 다른 주솟값을 가지더라도 이 object안에서는 같은 데이터를 공유받을 수 있습니다.
        @JvmStatic //Kotlin에는 Static이 없으므로 어노테이션을 이용하여 바로 접근
        fun createPet(user: User, request: CreatePetRequest): Pet {
            return Pet(
                user = user,
                name = request.name,
                gender = request.gender?.let { Gender.valueOf(it) },
                birthDate = request.birthDate,
                type = request.type
            )
        }
    }

    fun assignUser(newUser: User) {
        this.user?.pets?.remove(this)
        this.user = newUser
        newUser.pets.add(this)
    }

    fun updatePet(request: UpdatePetRequest) {
        if (request.isValidName()) {
            this.name = request.name
        }

        request.gender?.let {
            this.gender = Gender.valueOf(it)
        }

        request.type?.let {
            this.type = it
        }

        request.birthDate?.let {
            this.birthDate = it
        }
    }
}